server/src/main/scala/org/broadinstitute/dsde/workbench/hamm/api/HammRoutes.scala

```diff
@@ -0,0 +1,79 @@
+package org.broadinstitute.dsde.workbench.hamm.api
+
+import cats.effect._
+import io.circe.generic.auto._
+import org.broadinstitute.dsde.workbench.hamm.HammLogger
+import org.broadinstitute.dsde.workbench.hamm.auth.SamAuthProvider
+import org.broadinstitute.dsde.workbench.hamm.model.{HammException, JobId, WorkflowId}
+import org.broadinstitute.dsde.workbench.hamm.service._
+import org.http4s.Credentials.Token
+import org.http4s.circe.jsonEncoderOf
+import org.http4s.{AuthScheme, EntityEncoder, HttpRoutes, Request, Response, Status}
+import org.http4s.dsl.Http4sDsl
+import org.http4s.headers.{Allow, Authorization}
+import org.http4s.server.Router
+import org.http4s.server.middleware.Logger
+import org.http4s.syntax.kleisli._
+
+class HammRoutes(samDAO: SamAuthProvider, costService: CostService, statusService: StatusService)(implicit con: Concurrent[IO]) extends Http4sDsl[IO] with HammLogger {
+
+  // A Router can mount multiple services to prefixes.  The request is passed to the
+  //  service with the longest matching prefix.
+  def routes = Logger[IO](true, true)( Router[IO](
+    "/status" -> statusRoute,
+    "/api/cost/v1" -> costRoutes
+  ).orNotFound).mapF(handleException)
+
+
+  def statusRoute = HttpRoutes.of[IO] {
+    case GET -> Root =>
+      Ok( IO { statusService.status() } )
+    case _ -> Root =>
+      MethodNotAllowed(Allow(GET))
+  }
+
+
+  def costRoutes = HttpRoutes.of[IO] {
+    case request @ GET -> Root / "workflow" / workflowId =>
+      Ok(IO { costService.getWorkflowCost(extractToken(request), WorkflowId(workflowId)) })
+    case request @ GET -> Root / "job" / jobId =>
+      Ok(IO { costService.getJobCost(extractToken(request), JobId(jobId)) })
+    case _ -> Root =>
+      MethodNotAllowed(Allow(GET))
+  }
+
+
+
+
+
+
+  private def extractToken(request: Request[IO]): Token = {
```

Comment:
I would make this return an `Either[Throwable, Token]`, and at call site such as https://github.com/DataBiosphere/hamm/pull/5/files#diff-e77976a07ebc1116eacf55f51ac2b49bR38 something like

```
      for {
        token  <- IO.fromEither(extractToken(request))
        cost <- IO { costService.getWorkflowCost(token, WorkflowId(workflowId)) }
        resp <- Ok(cost)
      } yield resp
```

You don't have to squeezing everything in one line and wrap it in `OK`

Reply:
this feels like more code each time we make an authenticated call

Reply:
but it's a lot easier to see what could fail and what the errors are. By looking at `costService.getWorkflowCost(extractToken(request), WorkflowId(workflowId))`, you'll have to look closer to realize there's an `extractToken` call that could fail.

Reply:
I thought that it might be too much repetitive code, so I tried a different thing and added auth middleware